rllib/algorithms/dt/segmentation_buffer.py

```diff
@@ -0,0 +1,215 @@
+            logger.warning(
+                f"The maximum rollout length is {self.max_ep_len} but we tried to add a"
+                f"rollout of {episode.env_steps()} steps to the SegmentationBuffer. "
+                f"This could be due to incorrect data in the dataset (need fixing), "
+                f"or random shuffling that caused a duplicate rollout (no need to fix)."
+            )
+            episode = episode[: self.max_ep_len]
+
+        if len(self._buffer) < self.capacity:
+            self._buffer.append(episode)
+        else:
+            # TODO: replace proportional to episode length
+            replace_ind = np.random.randint(0, self.capacity)
+            self._buffer[replace_ind] = episode
+
+    def sample(self, batch_size: int) -> SampleBatch:
+        """Sample segments from the buffer.
+
+        Args:
+            batch_size: number of segments to sample.
+
+        Returns:
+            SampleBatch of segments with keys and shape {
+                OBS: [batch_size, max_seq_len, obs_dim],
+                ACTIONS: [batch_size, max_seq_len, act_dim],
+                RETURNS_TO_GO: [batch_size, max_seq_len + 1, 1],
+                T: [batch_size, max_seq_len],
+                ATTENTION_MASKS: [batch_size, max_seq_len],
+            }
+        """
+        samples = [self._sample_single() for _ in range(batch_size)]
+        return concat_samples(samples)
+
+    def _sample_single(self) -> SampleBatch:
+        # TODO: sample proportional to episode length
+        # Sample a random episode from the buffer and then sample a random
+        # segment from that episode.
+        buffer_ind = np.random.randint(0, len(self._buffer))
+        episode = self._buffer[buffer_ind]
+        ep_len = episode[SampleBatch.OBS].shape[0]
+        # This offset accounts for either the normal case when max_seq_len is shorter
+        # than the episode length, but also occasionally an episode will be shorter
+        # than the context length (max_seq_len).
+        offset = min(self.max_seq_len, ep_len)
+        # We allow si to be negative (for now) because we want segments that only
+        # contains the first few transitions (and padd the rest),
+        # for example [0, 0, 0, 0, 0, 0, R0, s0, a0].
+        si = np.random.randint(-offset + 1, ep_len - offset + 1)
+        ei = si + offset
+        # but for actual segmenting we don't want starting index to be negative.
+        si = max(si, 0)
+
+        obs = episode[SampleBatch.OBS][si:ei]
+        actions = episode[SampleBatch.ACTIONS][si:ei]
+        # Note that returns-to-go needs an extra elem as the target for the last action.
+        returns_to_go = episode[SampleBatch.RETURNS_TO_GO][si : ei + 1].reshape(-1, 1)
+
+        length = obs.shape[0]
+        timesteps = np.arange(si, si + length, dtype=np.int32)
+        masks = np.ones(length, dtype=returns_to_go.dtype)
+
+        # Back pad returns-to-go with 0 if at end of rollout.
```

Comment:
I don't think this works. I tried it before, and looked through the SampleBatch code, but basically SampleBatch does not support varied length arrays when concatenating multiple things together (when doing split_by_episode it will loop through the elements and slice which requires all keys to have the same size). This is calculated in postprocess_trajectory. 

If this is a really big issue, then what I could do is have postprocess_trajectory add a batch dimension to the episode, so that the batch is now a seq len 1 batch, where each element is the entire episode. I actually kinda like this idea because then we won't need to add use_dones=True at all, but that also requires redesigns of a few different places that use split_by_episodes in the original code base.

Reply:
Addressed issue with in person discussion. More to do with the code being unclear.

Reply:
Done!